AK/String.cpp

```diff
@@ -0,0 +1,338 @@
+}
+
+ErrorOr<NonnullRefPtr<StringData>> StringData::create_uninitialized(size_t byte_count, u8*& buffer)
+{
+    VERIFY(byte_count);
+    void* slot = malloc(allocation_size_for_string_data(byte_count));
+    if (!slot) {
+        return Error::from_errno(ENOMEM);
+    }
+    auto new_string_data = adopt_ref(*new (slot) StringData(byte_count));
+    buffer = const_cast<u8*>(new_string_data->bytes().data());
+    return new_string_data;
+}
+
+ErrorOr<NonnullRefPtr<StringData>> StringData::from_utf8(char const* utf8_data, size_t byte_count)
+{
+    // Strings of MAX_SHORT_STRING_LENGTH bytes or less should be handled by the String short string optimization.
+    VERIFY(byte_count > String::MAX_SHORT_STRING_LENGTH);
+
+    VERIFY(utf8_data);
+    u8* buffer = nullptr;
+    auto new_string_data = TRY(create_uninitialized(byte_count, buffer));
+    memcpy(buffer, utf8_data, byte_count * sizeof(char));
+    return new_string_data;
+}
+
+ErrorOr<NonnullRefPtr<StringData>> StringData::create_substring(StringData const& superstring, size_t start, size_t byte_count)
+{
+    // Strings of MAX_SHORT_STRING_LENGTH bytes or less should be handled by the String short string optimization.
+    VERIFY(byte_count > String::MAX_SHORT_STRING_LENGTH);
+
+    void* slot = malloc(sizeof(StringData) + sizeof(StringData::SubstringData));
+    if (!slot) {
+        return Error::from_errno(ENOMEM);
+    }
+    return adopt_ref(*new (slot) StringData(superstring, start, byte_count));
+}
+
+void StringData::compute_hash() const
+{
+    auto bytes = this->bytes();
+    if (bytes.size() == 0)
+        m_hash = 0;
+    else
+        m_hash = string_hash(reinterpret_cast<char const*>(bytes.data()), bytes.size());
+    m_has_hash = true;
+}
+
+Utf8View String::code_points() const
+{
+    return Utf8View(bytes_as_string_view());
+}
+
+ErrorOr<void> Formatter<String>::format(FormatBuilder& builder, String const& utf8_string)
+{
+    return Formatter<StringView>::format(builder, utf8_string.bytes_as_string_view());
+}
+
+ErrorOr<String> String::replace(StringView needle, StringView replacement, ReplaceMode replace_mode) const
+{
+    return StringUtils::replace(*this, needle, replacement, replace_mode);
+}
+
+bool String::is_short_string() const
+{
+    return reinterpret_cast<uintptr_t>(m_data) & 1;
+}
+
+ReadonlyBytes String::ShortString::bytes() const
+{
+    return { storage, byte_count() };
+}
+
+size_t String::ShortString::byte_count() const
+{
+    return byte_count_and_type >> 1;
+}
+
+unsigned Traits<String>::hash(String const& string)
+{
+    return string.hash();
+}
+
+DeprecatedString String::to_deprecated_string() const
+{
+    return DeprecatedString(bytes_as_string_view());
+}
+
+ErrorOr<String> String::from_deprecated_string(DeprecatedString const& deprecated_string)
+{
+    Utf8View view(deprecated_string);
+    if (!view.validate())
+        return Error::from_string_literal("String::from_deprecated_string: Input was not valid UTF-8");
```

Comment:
Neither `StringData::from_utf8()` nor `String::from_utf8()` seems to guarantee the validity of the provided UTF-8 data, unless I'm missing something. Should we really do this (only) here then?

Reply:
As in, if the new string is not happy to fully commit to UTF-8 validity, it should not assume that from the old string either.

Reply:
Let's commit.